flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/RemoteStreamEnvironment.java

```diff
@@ -169,39 +170,46 @@ public RemoteStreamEnvironment(String host, int port, Configuration clientConfig
 		}
 	}
 
-	@Override
-	public JobExecutionResult execute(String jobName) throws ProgramInvocationException {
-		StreamGraph streamGraph = getStreamGraph();
-		streamGraph.setJobName(jobName);
-		transformations.clear();
-		return executeRemotely(streamGraph, jarFiles);
+	protected List<URL> getJarFiles() throws ProgramInvocationException {
+		return jarFiles;
 	}
 
 	/**
 	 * Executes the remote job.
 	 *
-	 * @param streamGraph
-	 *            Stream Graph to execute
+	 * @param streamExecutionEnvironment
+	 *            Execution Environment with Stream Graph to execute
 	 * @param jarFiles
 	 * 			  List of jar file URLs to ship to the cluster
 	 * @return The result of the job execution, containing elapsed time and accumulators.
 	 */
-	protected JobExecutionResult executeRemotely(StreamGraph streamGraph, List<URL> jarFiles) throws ProgramInvocationException {
+	public static JobExecutionResult executeRemotely(StreamExecutionEnvironment streamExecutionEnvironment,
```

Comment:
I agree that we can't do this. The method is `protected` but it would still break subclasses relying on the `@Public` contract.

Reply:
Question: Why don't we pass the required dependencies directly here? i.e. `StreamGraph` and `ExecutionConfig`

IMHO there is no need to pass the entire environment.

Reply:
Using the environment is the simplest option and hides several internal classes (like the StreamGraph) that we shouldn't expose to users. Effectively this serves as an extension to the `RemoteStreamEnvironment` without baking it into a `RemoteStreamEnvironment` instance itself, which is fair imo.

IMO any use-case where usage of the StreamGraph is appropriate (for which tbh there are none as the JobGraph would be a better choice anyway) is already a pretty low-level thing in which case you can work directly against a ClusterClient.

Reply:
StreamGraph is exposed anyway because there is `StreamExecutionEnvironment.getStreamGraph()` though I see it is marked `@Internal`. I'm fine also with passing the environment.

>IMO any use-case where usage of the StreamGraph is appropriate (for which tbh there are none as the JobGraph would be a better choice anyway) is already a pretty low-level thing in which case you can work directly against a ClusterClient.

So you are suggesting to move the code here to `ClusterClient`? I think that would make sense.

Reply:
This cannot be moved to the ClusterClient since it relies on Streaming API classes (StreamExecutionEnvironment, or with your suggestion the StreamGraph) that aren't available in `flink-clients`. Hence why there we are only working on the `JobGraph`. And we can't just add a dependency since `flink-streaming-java` already depends on `flink-clients`.

Reply:
If we want to retain / put back the existing `protected JobExecutionResult executeRemotely`, then we would need to pass around `StreamGraph` instead of the environment. There are then other bits and pieces that also need to be passed around, such as the class loader. That's a bit difficult to see unless you actually change the code.

That protected method should ideally not be referenced outside of Flink, but as it happens we just used it to bridge the gap in Beam.. 

Regarding the static method, it isn't ideal to have this in the public API either, but that's the purpose of the change.. Every change to the parameter list becomes an incompatible change and may cause pain going forward. 

I have also added the ability to set the savepoint option on the environment, which seems more user friendly. The minimum change btw is here, in case we want to reconsider the route we are taking here: https://github.com/apache/flink/compare/master...tweise:FLINK-11048.savepointRestore.min?expand=1

Reply:
Agreed on the public method contract. I rearranged the code to keep the existing method as is.

The setter is an alternative option for users that already have the remote environment instance (instead of having to deal with an 8 parameter static method). I would like to keep this as is.

Reply:
Looks good but please see my comment above concerning the setter.

Reply:
Do we still need this new `static` method now that we have the constructor?

Reply:
Yes. We want the ability to execute a job remotely regardless of the type of environment.